File: July 2024/14-07-2024/Segregate 0s and 1s.cpp
//{ Driver Code Starts
// Initial template for C++
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:
    void segregate0and1(vector<int> &arr) {
//Your Code Start
        int l = 0;
        int r = arr.size() - 1;
         while (l < r) {
            while (arr[l] == 0 && l< r) {
                l++;
            }
            while (arr[r] == 1 && l < r) {
                r--;
            }
            if (l < r) {
                std::swap(arr[l], arr[r]);
                l++;
                r--;
            }
        }
        //Your Code End
    }
};

//{ Driver Code Starts.
int main() {
    int t;

    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        string input;
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }
        int n = arr.size();
        Solution ob;
        ob.segregate0and1(arr);
        for (int i = 0; i < n; i++) {
            cout << arr[i] << " ";
        }
        cout << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: July 2024/14-07-2024/Segregate 0s and 1s.java
//{ Driver Code Starts


// Initial Template for Java

import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;

            new Solution().segregate0and1(arr);
            for (int i = 0; i < array.size(); i++) {
                System.out.print(arr[i] + " ");
            }
            System.out.println();
            // System.out.println("~");
        }
    }
}

// } Driver Code Ends


// User function Template for Java

class Solution {
    void segregate0and1(int[] arr) {
   //Your Code Start
        int l = 0;
        int r = arr.length -1;
        while (l<r){

            while(arr[l] == 0 && l<r){
                l += 1; 
            }
            
            while(arr[r]==1 && l<r){
                r -= 1;
            }
       
            if (l < r){
              int temp = arr[l];
                arr[l] = arr[r];
                arr[r] = temp;
                l += 1;
                r -= 1;
            }
      
    }
      //Your Code End
        
    }

}
